### PR TITLE
Fix reduction='none' in poisson_nll_loss

### LINKED ISSUE #17347: F.poisson_nll_loss ignores  reduction='none'
Hi!
Using PyTorch 1.0.1, it seems like  F.poisson_nll_loss ignores the `reduction='none'`:

```
import torch
import torch.nn.functional as F
input = torch.tensor([0.5, 1.5, 2.5])
target = torch.tensor([1., 2., 3.])
print(F.poisson_nll_loss(input, target, reduction='none'))
```
gives
```tensor(7.3129)```

i.e.,  it ignores the `reduction='none'` argument.

### PR DESCRIPTION
**Pull request by vishwakftw**
Changelog:
- Modify `if` to `elif` in reduction mode comparison
- Add error checking for reduction mode

Test plan:
- Added tests in `test_nn.py`

Fixes https://github.com/pytorch/pytorch/issues/17347.

### CODE REVIEW

**Review by zou3519 — APPROVED** [review]
thank you!

File: torch/nn/functional.py [hunk 1]
```diff
@@ -1909,10 +1909,13 @@ def poisson_nll_loss(input, target, log_input=True, full=False, size_average=Non
...
     if reduction == 'none':
         ret = loss
-    if reduction == 'mean':
+    elif reduction == 'mean':
         ret = torch.mean(loss)
-    else:
+    elif reduction == 'sum':
         ret = torch.sum(loss)
+    else:
+        ret = input
```
Comment (zou3519) [inline torch/nn/functional.py]: nit: We don't need this anymore, right?

**Comment by zou3519** [comment]
I think if you rebase onto master all of those CI failures will go away (they look like infra failures)

File: torch/nn/functional.py (hunk 1 above)
Comment (vishwakftw) [reply torch/nn/functional.py]: I think JIT will throw up an error, because we wouldn't be true to the type annotation.

Comment (zou3519) [reply torch/nn/functional.py]: Interesting! That's fine then, thanks for figuring that out

**Comment by vishwakftw** [comment]
@zou3519 is there anything I have to do to get this landed?

**Comment by vishwakftw** [comment]
Thank you!

**Comment by zou3519** [comment]
@ezyang Do you think it is worth investing in better land failure notifications? For example, if the bot comments on the PR that the land failed.

I didn't notice that this PR didn't land until a few days after I hit the land button.

**Comment by ezyang** [comment]
My first suggestion is, try regularly checking your personal diff queue in internal Phabricator. That's supposed to give you signal about if lands are working or not; I need to do this anyway for diffs I work on. If that doesn't work, let's figure something out.

**Pull request closed without merging**